tests/exponential.rs

```diff
@@ -0,0 +1,402 @@
+/*!
+    Exponential functions from arithmetic
+!*/
+
+use num::rational::Ratio;
+use num::BigInt;
+use ruler::*;
+
+pub type Rational = Ratio<BigInt>;
+
+egg::define_language! {
+    pub enum Exponential {
+        // trig operators
+        "exp" = Exp(Id),
+        "log" = Log(Id),
+        "pow" = Pow([Id; 2]),
+        "sqrt" = Sqrt(Id),
+        "cbrt" = Cbrt(Id),
+
+        // arithmetic operators
+        "~" = Neg(Id),
+        "+" = Add([Id; 2]),
+        "-" = Sub([Id; 2]),
+        "*" = Mul([Id; 2]),
+        "/" = Div([Id; 2]),
+
+        // constants
+        Num(Rational),
+        Var(Symbol),
+    }
+}
+
+impl SynthLanguage for Exponential {
+    type Constant = Rational;
+
+    // cvec-less domain
+    fn eval<'a, F>(&'a self, _cvec_len: usize, _get_cvec: F) -> CVec<Self>
+    where
+        F: FnMut(&'a Id) -> &'a CVec<Self>,
+    {
+        vec![]
+    }
+
+    fn initialize_vars(egraph: &mut EGraph<Self, SynthAnalysis>, vars: &[String]) {
+        for v in vars {
+            let id = egraph.add(Exponential::Var(Symbol::from(v.clone())));
+
+            let l_id = egraph.add(Exponential::Log(id));
+            let el_id = egraph.add(Exponential::Exp(l_id));
+
+            let e_id = egraph.add(Exponential::Exp(id));
+            let le_id = egraph.add(Exponential::Log(e_id));
+
+            egraph.union(id, el_id);
+            egraph.union(id, le_id);
+            egraph.rebuild();
+        }
+    }
+
+    fn to_var(&self) -> Option<Symbol> {
+        if let Exponential::Var(sym) = self {
+            Some(*sym)
+        } else {
+            None
+        }
+    }
+
+    fn mk_var(sym: Symbol) -> Self {
+        Exponential::Var(sym)
+    }
+
+    fn is_constant(&self) -> bool {
+        matches!(self, Exponential::Num(_))
+    }
+
+    fn mk_constant(c: Self::Constant, _egraph: &mut EGraph<Self, SynthAnalysis>) -> Self {
+        Exponential::Num(c)
+    }
+
+    // no validation possible
+    fn validate(_lhs: &Pattern<Self>, _rhs: &Pattern<Self>) -> ValidationResult {
+        ValidationResult::Valid
+    }
+
+    fn is_rule_lifting() -> bool {
+        true
+    }
+
+    fn get_lifting_rules() -> enumo::Ruleset<Self> {
+        enumo::Ruleset::from_str_vec(&[
+            // definitions (denote)
```

Comment:
I'm still concerned about these in terms of the allowed/forbidden narrative, but it's fine for now